src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEFileEditHintReceiver.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.VS.Automation;
+using VSLangProj;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.TempPE
+{
+    /// <summary>
+    /// This class handles being notified whenever a source file is edited, in our case by a Single File Generator
+    /// </summary>
+    [Export(typeof(IProjectChangeHintReceiver))]
+    [ProjectChangeHintKind(ProjectChangeFileSystemEntityHint.EditedFileAsString)]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class TempPEFileEditHintReceiver : IProjectChangeHintReceiver
+    {
+        private readonly VSBuildManager _buildManager;
+        private readonly Lazy<ITempPEBuildManager> _tempPEBuildManager;
+
+        [ImportingConstructor]
+        public TempPEFileEditHintReceiver(
+            [Import(typeof(BuildManager))]VSBuildManager buildManager,
+            Lazy<ITempPEBuildManager> tempPEBuildManager)
+        {
+            _buildManager = buildManager;
+            _tempPEBuildManager = tempPEBuildManager;
+        }
+
+        public async Task HintedAsync(IImmutableDictionary<Guid, IImmutableSet<IProjectChangeHint>> hints)
+        {
+            // We don't want to be the one to realise the Lazy<ITempPEManager> so ignore changes in projects that don't need TempPE
+            if (!_buildManager.ProjectNeedsTempPE) return;
+
+            IImmutableSet<IProjectChangeHint> files = hints[ProjectChangeFileSystemEntityHint.EditedFile];
+            foreach (IProjectChangeFileHint hint in files)
+            {
+                foreach (string fileName in hint.Files)
+                {
+                    // The magic of MEF means that the instance we get from .Value is the same as the build managers instance, so it has all of the info it needs
```

Comment:
This isn't magic, this is as a result of shared instances. The couple here is I think needs to be removed - I don't it's BuildManager concept to know if TempPE is running.

Reply:
This is related to https://github.com/dotnet/project-system/pull/4349#discussion_r239280321 and will be removed with that change.